service/frontend/task_reachability.go

```diff
@@ -0,0 +1,299 @@
+}
+
+func (wh *WorkflowHandler) getBuildIdReachability(
+	ctx context.Context,
+	request buildIdReachabilityRequest,
+) (*taskqueuepb.BuildIdReachability, error) {
+	taskQueues := request.taskQueues
+	if len(taskQueues) == 0 {
+		// Namespace scope, fetch mapping from DB.
+		response, err := wh.matchingClient.GetBuildIdTaskQueueMapping(ctx, &matchingservice.GetBuildIdTaskQueueMappingRequest{
+			NamespaceId: request.namespace.ID().String(),
+			BuildId:     request.buildId,
+		})
+		if err != nil {
+			return nil, err
+		}
+		taskQueues = response.TaskQueues
+	}
+
+	numTaskQueuesToQuery := util.Min(len(taskQueues), wh.config.ReachabilityTaskQueueScanLimit())
+	taskQueuesToQuery, taskQueuesToSkip := taskQueues[:numTaskQueuesToQuery], taskQueues[numTaskQueuesToQuery:]
+
+	taskQueueReachability, err := util.MapConcurrent(taskQueuesToQuery, func(taskQueue string) (*taskqueuepb.TaskQueueReachability, error) {
+		compatibilityResponse, err := request.versionSetFetcher.fetchTaskQueueVersions(ctx, request.namespace, taskQueue)
+		if err != nil {
+			return nil, err
+		}
+		return wh.getTaskQueueReachability(ctx, taskQueueReachabilityRequest{
+			buildId:          request.buildId,
+			namespace:        request.namespace,
+			taskQueue:        taskQueue,
+			versionSets:      compatibilityResponse.GetResponse().GetMajorVersionSets(),
+			reachabilityType: request.reachabilityType,
+		})
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(taskQueuesToSkip) > 0 {
+		skippedTasksReachability := make([]*taskqueuepb.TaskQueueReachability, len(taskQueuesToSkip))
+		for i, taskQueue := range taskQueuesToSkip {
+			skippedTasksReachability[i] = &taskqueuepb.TaskQueueReachability{TaskQueue: taskQueue, Reachability: []enumspb.TaskReachability{enumspb.TASK_REACHABILITY_UNSPECIFIED}}
+		}
+		taskQueueReachability = append(taskQueueReachability, skippedTasksReachability...)
+	}
+	return &taskqueuepb.BuildIdReachability{BuildId: request.buildId, TaskQueueReachability: taskQueueReachability}, nil
+}
+
+type taskQueueReachabilityRequest struct {
+	buildId          string
+	taskQueue        string
+	namespace        *namespace.Namespace
+	versionSets      []*taskqueuepb.CompatibleVersionSet
+	reachabilityType enumspb.TaskReachability
+}
+
+// Get the reachability of a single build ID in a single task queue scope.
+func (wh *WorkflowHandler) getTaskQueueReachability(ctx context.Context, request taskQueueReachabilityRequest) (*taskqueuepb.TaskQueueReachability, error) {
+	taskQueueReachability := taskqueuepb.TaskQueueReachability{TaskQueue: request.taskQueue, Reachability: []enumspb.TaskReachability{}}
+
+	var isDefaultInQueue bool
+	var buildIdsFilter string
+
+	if request.buildId == "" {
+		// Query for the unversioned worker
+		isDefaultInQueue = len(request.versionSets) == 0
+		// Query workflows that have completed tasks marked with a sentinel "unversioned" search attribute.
+		buildIdsFilter = fmt.Sprintf("BuildIds = %q", common.UnversionedSearchAttribute)
```

Comment:
don't trust %q to use the same types of quote and escaping as our sql parser
```suggestion
		buildIdsFilter = fmt.Sprintf(`BuildIds = "%s"`, common.UnversionedSearchAttribute)
```
(this is a constant so no need to escape, but you could if you want)